pkg/framework/kind/type.cue

```diff
@@ -0,0 +1,111 @@
+//
+// Classes and objects in most programming languages are analogous:
+//   - #Kind is like a `class` keyword
+//   - Each declaration of #Kind is like a class declaration
+//   - Byte sequences are like arguments to the class constructor
+//   - Entities are like objects - what's returned from the constructor
+//
+// There are two fundamental varieties of Kinds: Raw and Structured.
+#Kind: #Raw | #Structured
+
+// properties shared between all kind varieties.
+_sharedKind: {
+	// name is the canonical, machine-friendly name of a Kind.
+	//
+	// name is often used in generated code intended to be parsed be parsed or used
+	// by machines, so it follows the relatively strict DNS label naming standard,
+	// as defined in RFC 1123:
+	//  - Contain at most 63 characters
+	//  - Contain only lowercase alphanumeric characters or '-'
+	//  - Start with an alphabetic character
+	//  - End with an alphanumeric character
+	name: =~"^([a-z][a-z0-9-]{0,61}[a-z0-9])$"
+
+	// TODO need some fields related to other names, human names - pluralization, capitalization
+	// it's all necessary for conversion to CRDs
+
+	// maturity indicates the how far a given kind declaration is in its initial
+	// journey towards what might otherwise be referred to as 1.0.
+	maturity: *"committed" | "synchronized" | "mature"
+
+	// The entity system itself is not mature enough yet for any individual
+	// kind to be called "mature."
+	maturity: *"committed" | "synchronized"  // TODO unclear if we want maturity for raw kinds
+
+	form: "structured" | "raw"
+}
+
+// Raw is a variety of Kind that specifies a type for a raw file,
+// like an svg or parquet file. Grafana mostly acts as asset storage for raw
+// types: the byte sequence is a black box to Grafana, and type is determined
+// through metadata such as file extension.
+#Raw: {
+	_sharedKind
+	form: "raw"
+	extensions?: [...string]
+	// need
+	// - sanitize function
+	// - get summary
+}
+
+// Structured is a variety of Kind where a schema specifies
+// validity rules for the byte sequence. These represent all the conventional
+// types and functional resources in Grafana, such as dashboards and datasources.
+//
+// Structured types may be defined either by Grafana itself (core types), or by
+// plugins. Plugin-defined types have a slightly reduced set of capabilities,
+// due to the constraints imposed by them being run in separate processes, and
+// the risks arising from executing code from potentially untrusted third parties.
+#Structured: {
+	_sharedKind
+	#CoreStructured | #CustomStructured
+	form:     "structured"
+
+	lineage:  thema.#Lineage
+
+	// currentVersion is the syntactic version of the lineage's latest schema.
+	currentVersion: thema.#SyntacticVersion & (thema.#LatestVersion & { lin: lineage }).out
+}
+
+#CoreStructured: {
+	#Structured
+}
+
+// NOTE not a definition b/c apps/plugins could be written against a later Grafana version
+CustomStructured: {
+	#Structured
+}
```

Comment:
This is a bit of conventional CUE practice - [definitions](https://cuelang.org/docs/tutorials/tour/intro/schema/) (`#`-led names) are normally interpreted as closed, meaning that an error occurs if they're unified with data that has fields not specified in the definition.

This is a forward compatibility thing. So, for stuff that's expected to be used solely in Grafana core, that's fine - those things are part of the same repo, and therefore have the same/a predictable version vs. `#CoreStructured`. 

However, externally-defined plugins do not. If we introduce a new metadata field on `#CustomStructured` (with the leading hash) in, say, Grafana 10.5, and a plugin includes that field in one of their type definitions, then a running Grafana 10.4 instance would reject that plugin's type as an excess field because it's not in `#CustomStructured`.

For some cases, that may be what we want - but it should be managed explicitly with a minimum grafana version number somewhere, not implicitly through a schema metadata validation failure.